tests/hotplug/instancetype.go

```diff
@@ -182,4 +184,111 @@ var _ = Describe("[sig-compute]Instance Type and Preference Hotplug", decorators
 		Entry("without maxGuest and maxSockets defined", false),
 	)
+
+	DescribeTable("should reject live update when preference requirements are no longer met - bug #14595",
+		func(
+			createVM func(*virtv1.VirtualMachineInstance, string) (*virtv1.VirtualMachine, error),
+			updateVM func(*virtv1.VirtualMachine) (*virtv1.VirtualMachine, error),
+			errMatcher gomegatypes.GomegaMatcher,
+		) {
+			vmi := libvmifact.NewAlpine(
+				libnet.WithMasqueradeNetworking(),
+				libvmi.WithResourceMemory("1Gi"),
+			)
+			vmi.Namespace = testsuite.GetTestNamespace(vmi)
+
+			preference := builder.NewPreference()
+			preference.Spec.Requirements = &instancetypev1beta1.PreferenceRequirements{
+				Memory: &instancetypev1beta1.MemoryPreferenceRequirement{
+					Guest: resource.MustParse("1Gi"),
+				},
+			}
+			preference, err := virtClient.VirtualMachinePreference(vmi.Namespace).Create(
+				context.Background(), preference, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			vm, err := createVM(vmi, preference.Name)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = updateVM(vm)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(errMatcher)
+		},
+		Entry("by VirtualMachine resource requests",
+			func(vmi *virtv1.VirtualMachineInstance, preferenceName string) (*virtv1.VirtualMachine, error) {
+				return virtClient.VirtualMachine(vmi.Namespace).Create(
+					context.Background(),
+					libvmi.NewVirtualMachine(vmi, libvmi.WithPreference(preferenceName)),
+					metav1.CreateOptions{},
+				)
+			},
+			func(vm *virtv1.VirtualMachine) (*virtv1.VirtualMachine, error) {
+				newMemory := vm.Spec.Template.Spec.Domain.Resources.Requests.Memory().DeepCopy()
+				newMemory.Sub(resource.MustParse("512Mi"))
+
+				const memoryPath = "/spec/template/spec/domain/resources/requests/memory"
+				patches := patch.New(
+					patch.WithTest(memoryPath, vm.Spec.Template.Spec.Domain.Resources.Requests.Memory().String()),
+					patch.WithReplace(memoryPath, newMemory.String()),
+				)
+				patchData, err := patches.GeneratePayload()
+				Expect(err).NotTo(HaveOccurred())
+				return virtClient.VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, patchData, metav1.PatchOptions{})
+			},
+			ContainSubstring("failure checking preference requirements: insufficient Memory resources of 512Mi provided by VirtualMachine, preference requires 1Gi"),
+		),
+		Entry("by instance type resource requests",
+			func(vmi *virtv1.VirtualMachineInstance, preferenceName string) (*virtv1.VirtualMachine, error) {
+				origInstancetype, err := virtClient.VirtualMachineInstancetype(vmi.Namespace).Create(
+					context.Background(),
+					builder.NewInstancetypeFromVMI(vmi),
+					metav1.CreateOptions{},
+				)
+				Expect(err).ToNot(HaveOccurred())
+
+				vm, err := virtClient.VirtualMachine(vmi.Namespace).Create(
+					context.Background(),
+					libvmi.NewVirtualMachine(
+						vmi,
```

Comment:
Will there be no conflict with the VMI's `libvmi.WithResourceMemory("1Gi")`?

Reply:
Yeah, `WithInstancetype` calls `resourcesRemovedFromVMI` to handle this:

https://github.com/kubevirt/kubevirt/blob/21ce93e095e3f183d939422dfe3cb661ae7525d4/pkg/libvmi/vm.go#L137-L145

https://github.com/kubevirt/kubevirt/blob/21ce93e095e3f183d939422dfe3cb661ae7525d4/pkg/libvmi/vm.go#L101-L105

Reply:
Interesting to know, I would have not expected `libvmi` to contain such logic.